File: 20-valid-parentheses/20-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        match = {")" : "(", "]" : "[", "}" : "{"}
        
        for char in s:
            if char in match: ##closing
                top = stack.pop() if stack else '*' ##dummy
                if match[char] != top: return False
            else:
                stack.append(char)
                
        return not stack ##len(stack) == 0
        
        

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> match = new HashMap(); 
        match.put(')', '(');
        match.put(']', '[');
        match.put('}', '{');
        Stack<Character> stack = new Stack();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (match.containsKey(c)) {
                if (stack.isEmpty()) {
                    return false;
                }
                else {
                    char top = stack.pop();
                    if (match.get(c) != top) {return false;}
                }
            }
            else {
                stack.push(c);
            }
        }
        
        return stack.isEmpty();
    }
}